routing/rt/rt.go

```diff
@@ -0,0 +1,324 @@
+// Copyright (c) 2016 Bitfury Group Limited
+// Distributed under the MIT software license, see the accompanying
+// file LICENSE or http://www.opensource.org/licenses/mit-license.php
+
+package rt
+
+import (
+	"encoding/gob"
+	"fmt"
+	"github.com/lightningnetwork/lnd/routing/rt/graph"
+	"io"
+)
+
+type OperationType uint32
+const (
+	ADD_CHANNEL OperationType = iota
+	REMOVE_CHANNEL
+)
+
+func (t OperationType) String()string {
+	switch t {
+	case ADD_CHANNEL:
+		return "<ADD_CHANNEL>"
+	case REMOVE_CHANNEL:
+		return "<REMOVE_CHANNEL>"
+	default:
+		return "<UNKNOWN_TYPE>"
+	}
+}
+type Channel struct {
+	Id1, Id2 graph.ID
+	EdgeID   graph.EdgeID
+	Info     *graph.ChannelInfo
+}
+
+type ChannelOperation struct {
+	Id1, Id2  graph.ID
+	EdgeID    graph.EdgeID
+	Info      *graph.ChannelInfo
+	Operation OperationType // One of ADD_CHANNEL, REMOVE_CHANNEL
+}
+
+func (c ChannelOperation) String () string {
+	return fmt.Sprintf(
+		"ChannelOperation[%v %v %v %v %v]",
+		c.Id1,
+		c.Id2,
+		c.EdgeID,
+		c.Info,
+		c.Operation,
+	)
+}
+
+func (op *ChannelOperation) Copy() ChannelOperation {
+	return ChannelOperation{
+		Id1:       op.Id1,
+		Id2:       op.Id2,
+		EdgeID:    op.EdgeID,
+		Info:      op.Info.Copy(),
+		Operation: op.Operation,
+	}
+}
+
+type DifferenceBuffer []ChannelOperation
+
+func (d DifferenceBuffer) String() string {
+	s := ""
+	for i:=0; i<len(d); i++ {
+		s += fmt.Sprintf("  %v\n", d[i])
+	}
+	s1 := ""
+	if len(d) > 0 {
+		s1 = "\n"
+	}
+	return fmt.Sprintf("DifferenceBuffer[%v%v]", s1, s)
+}
+
+func NewDifferenceBuffer() *DifferenceBuffer {
+	d := make(DifferenceBuffer, 0)
+	return &d
+}
+
+func (diffBuff *DifferenceBuffer) IsEmpty() bool {
+	return len(*diffBuff) == 0
```

Comment:
Actually not
routing/rt/rt.go:91: invalid argument diffBuff (type *DifferenceBuffer) for len